File: Leetcode/0662 Maximum Width of Binary Tree.py
class Solution:
    def widthOfBinaryTree(self, root):
        queue = deque([(root, 0)])
        max_width = 0
        while queue:
            width = queue[-1][1] - queue[0][1] + 1
            max_width = max(max_width, width)
            for _ in range(len(queue)):
                node, index = queue.popleft()
                if node.left is not None:
                    queue.append((node.left, 2 * index + 1))
                if node.right is not None:
                    queue.append((node.right, 2 * index + 2))
        return max_width

File: Leetcode/0662 Maximum Width of Binary Tree.cpp
class Solution {
public:
    int widthOfBinaryTree(TreeNode* root) {
        deque<pair<TreeNode*, int>> queue {{root, 0}};
        int maxWidth = 0;
        while (not queue.empty()) {
            int width = (long long) queue.back().second - queue.front().second + 1;
            maxWidth = max(maxWidth, width);
            int size = queue.size();
            for (int i = 0;i < size; i++) {
                auto [node, index] = queue.front();
                queue.pop_front();
                long long leftIndex = (long long) 2 * index;
                if (node->left != nullptr) {
                    queue.push_back({node->left, leftIndex});
                }
                if (node->right != nullptr) {
                    queue.push_back({node->right, leftIndex + 1});
                }
            } 
        }
        return maxWidth;
    }
};
